### DISCUSSION TITLE
Cannot run appium inspector on real Android device

### DISCUSSION THREAD

**Discussion by murigregori** [appium/appium#20837] [Q&A]
Hello, I'd like to ask for help on running the appium inspector on an Android device. It seems that the appium inspector always gets a request timeout error whenever i try to open an app on my device via appium inspector. It was working well before and unfortunately, I wasnt able to work on it for a long time and now, I am not able to run appium inspector on my android device despite updating all drivers, appium server, and appium inspector as well. These are my capabilities when trying to run: 

 "platformName": "Android",
    "appium:deviceName": "220333QAG",
    "appium:udid": "9a424a1b",
    "appium:platformVersion": "13",
    "appium:appPackage": "com.paymaya.staging",
    "appium:appActivity": "com.paymaya.ui.login.view.activity.impl.LauncherActivity",
    "appium:automationName": "UiAutomator2",
    "appium:uiautomator2ServerInstallTimeout": 40000,
    "appium:adbExecTimeout": 40000

and here is a sample log of the issue encountered

`[AppiumDriver@30d3] Calling AppiumDriver.createSession() with args: [{"appium:deviceName":"220333QAG","appium:udid":"9a424a1b","platformName":"Android","appium:platformVersion":"13","appium:appPackage":"com.paymaya.staging","appium:appActivity":"com.paymaya.ui.login.view.activity.impl.LauncherActivity","appium:automationName":"UiAutomator2","appium:uiautomator2ServerInstallTimeout":40000,"appium:adbExecTimeout":40000,"appium:ensureWebviewsHavePages":true,"appium:nativeWebScreenshot":true,"appium:newCommandTimeout":3600,"appium:connectHardwareKeyboard":true},null,{"alwaysMatch":{"appium:deviceName":"220333QAG","appium:udid":"9a424a1b","platformName":"Android","appium:platformVersion":"13","appium:appPackage":"com.paymaya.staging","appium:appActivity":"com.paymaya.ui.login.view.activity.impl.LauncherActivity","appium:automationName":"UiAutomator2","appium:uiautomator2ServerInstallTimeout":40000,"appium:adbExecTimeout":40000,"appium:ensureWebviewsHavePages":true,"appium:nativeWebScreenshot":true,"appium:newCommandTimeout":3600,"appium:connectHardwareKeyboard":true},"firstMatc...
[AppiumDriver@30d3] Event 'newSessionRequested' logged at 1734076441027 (15:54:01 GMT+0800 (Philippine Standard Time))
[Appium] Attempting to find matching driver for automationName 'UiAutomator2' and platformName 'Android'
[Appium] The 'uiautomator2' driver was installed and matched caps.
[Appium] Will require it at /Users/meryl.gregori/.appium/node_modules/appium-uiautomator2-driver
[Appium] Requiring driver at /Users/meryl.gregori/.appium/node_modules/appium-uiautomator2-driver/build/index.js
[AppiumDriver@30d3] Appium v2.13.1 creating new AndroidUiautomator2Driver (v3.9.3) session
[AppiumDriver@30d3] Checking BaseDriver versions for Appium and AndroidUiautomator2Driver
[AppiumDriver@30d3] Appium's BaseDriver version is 9.13.1
[AppiumDriver@30d3] AndroidUiautomator2Driver's BaseDriver version is 9.13.1
[AndroidUiautomator2Driver@6f0d] 
[AndroidUiautomator2Driver@6f0d] Creating session with W3C capabilities: {
  "alwaysMatch": {
    "platformName": "Android",
    "appium:deviceName": "220333QAG",
    "appium:udid": "9a424a1b",
    "appium:platformVersion": "13",
    "appium:appPackage": "com.paymaya.staging",
    "appium:appActivity": "com.paymaya.ui.login.view.activity.impl.LauncherActivity",
    "appium:automationName": "UiAutomator2",
    "appium:uiautomator2ServerInstallTimeout": 40000,
    "appium:adbExecTimeout": 40000,
    "appium:ensureWebviewsHavePages": true,
    "appium:nativeWebScreenshot": true,
    "appium:newCommandTimeout": 3600,
    "appium:connectHardwareKeyboard": true
  },
  "firstMatch": [
    {}
  ]
}
[AndroidUiautomator2Driver@6f0d] The following provided capabilities were not recognized by this driver:
[AndroidUiautomator2Driver@6f0d]   connectHardwareKeyboard
[f34e677f][AndroidUiautomator2Driver@6f0d] Session created with session id: f34e677f-6ced-4e30-a4c3-72c3fddccd98
[f34e677f][ADB] Using 'adb' from '/Users/meryl.gregori/Library/Android/sdk/platform-tools/adb'
[f34e677f][ADB] Running '/Users/meryl.gregori/Library/Android/sdk/platform-tools/adb -P 5037 start-server'
[f34e677f][AndroidUiautomator2Driver@6f0d] Retrieving device list
[f34e677f][ADB] Trying to find connected Android devices
[f34e677f][ADB] Getting connected devices
[f34e677f][ADB] Connected devices: [{"udid":"9a424a1b","state":"device"}]
[f34e677f][AndroidUiautomator2Driver@6f0d] Using device: 9a424a1b
[f34e677f][ADB] Using 'adb' from '/Users/meryl.gregori/Library/Android/sdk/platform-tools/adb'
[f34e677f][ADB] Running '/Users/meryl.gregori/Library/Android/sdk/platform-tools/adb -P 5037 start-server'
[f34e677f][ADB] Setting device id to 9a424a1b
[f34e677f][AndroidUiautomator2Driver@6f0d] Starting 'com.paymaya.staging' directly on the device
[f34e677f][ADB] Running '/Users/meryl.gregori/Library/Android/sdk/platform-tools/adb -P 5037 -s 9a424a1b shell getprop ro.build.version.sdk'
[f34e677f][ADB] Current device property 'ro.build.version.sdk': 33
[f34e677f][ADB] Getting device platform version
[f34e677f][ADB] Running '/Users/meryl.gregori/Library/Android/sdk/platform-tools/adb -P 5037 -s 9a424a1b shell getprop ro.build.version.release'
[f34e677f][ADB] Current device property 'ro.build.version.release': 13
[f34e677f][ADB] Device API level: 33
[f34e677f][AndroidUiautomator2Driver@6f0d] Relaxing hidden api policy
[f34e677f][ADB] Running '/Users/meryl.gregori/Library/Android/sdk/platform-tools/adb -P 5037 -s 9a424a1b shell 'settings put global hidden_api_policy_pre_p_apps 1;settings put global hidden_api_policy_p_apps 1;settings put global hidden_api_policy 1''
[f34e677f][AndroidUiautomator2Driver@6f0d] No app sent in, not parsing package/activity
[f34e677f][AndroidUiautomator2Driver@6f0d] Pushing settings apk to the device...
[f34e677f][ADB] Getting package info for 'io.appium.settings'
[f34e677f][ADB] Running '/Users/meryl.gregori/Library/Android/sdk/platform-tools/adb -P 5037 -s 9a424a1b shell dumpsys package io.appium.settings'
[f34e677f][ADB] App '/Users/meryl.gregori/.appium/node_modules/appium-uiautomator2-driver/node_modules/io.appium.settings/apks/settings_apk-debug.apk' is not installed
[f34e677f][ADB] Installing '/Users/meryl.gregori/.appium/node_modules/appium-uiautomator2-driver/node_modules/io.appium.settings/apks/settings_apk-debug.apk'
[f34e677f][ADB] Running '/Users/meryl.gregori/Library/Android/sdk/platform-tools/adb -P 5037 -s 9a424a1b features'
[f34e677f][ADB] The application at '/Users/meryl.gregori/.appium/node_modules/appium-uiautomator2-driver/node_modules/io.appium.settings/apks/settings_apk-debug.apk' will not be cached, because the device under test has confirmed the support of streamed installs
[f34e677f][ADB] Running '/Users/meryl.gregori/Library/Android/sdk/platform-tools/adb -P 5037 -s 9a424a1b install -g /Users/meryl.gregori/.appium/node_modules/appium-uiautomator2-driver/node_modules/io.appium.settings/apks/settings_apk-debug.apk'
[f34e677f][ADB] Getting package info for 'io.appium.settings'
[f34e677f][ADB] Running '/Users/meryl.gregori/Library/Android/sdk/platform-tools/adb -P 5037 -s 9a424a1b shell dumpsys package io.appium.settings'
[f34e677f][ADB] App '/Users/meryl.gregori/.appium/node_modules/appium-uiautomator2-driver/node_modules/io.appium.settings/apks/settings_apk-debug.apk' is not installed
[f34e677f][ADB] Installing '/Users/meryl.gregori/.appium/node_modules/appium-uiautomator2-driver/node_modules/io.appium.settings/apks/settings_apk-debug.apk'
[f34e677f][ADB] The application at '/Users/meryl.gregori/.appium/node_modules/appium-uiautomator2-driver/node_modules/io.appium.settings/apks/settings_apk-debug.apk' will not be cached, because the device under test has confirmed the support of streamed installs
[f34e677f][ADB] Running '/Users/meryl.gregori/Library/Android/sdk/platform-tools/adb -P 5037 -s 9a424a1b install -g /Users/meryl.gregori/.appium/node_modules/appium-uiautomator2-driver/node_modules/io.appium.settings/apks/settings_apk-debug.apk'
[f34e677f][ADB] Getting package info for 'io.appium.settings'
[f34e677f][ADB] Running '/Users/meryl.gregori/Library/Android/sdk/platform-tools/adb -P 5037 -s 9a424a1b shell dumpsys package io.appium.settings'
[f34e677f][ADB] App '/Users/meryl.gregori/.appium/node_modules/appium-uiautomator2-driver/node_modules/io.appium.settings/apks/settings_apk-debug.apk' is not installed
[f34e677f][ADB] Installing '/Users/meryl.gregori/.appium/node_modules/appium-uiautomator2-driver/node_modules/io.appium.settings/apks/settings_apk-debug.apk'
[f34e677f][ADB] The application at '/Users/meryl.gregori/.appium/node_modules/appium-uiautomator2-driver/node_modules/io.appium.settings/apks/settings_apk-debug.apk' will not be cached, because the device under test has confirmed the support of streamed installs
[f34e677f][ADB] Running '/Users/meryl.gregori/Library/Android/sdk/platform-tools/adb -P 5037 -s 9a424a1b install -g /Users/meryl.gregori/.appium/node_modules/appium-uiautomator2-driver/node_modules/io.appium.settings/apks/settings_apk-debug.apk'
[f34e677f][AndroidUiautomator2Driver@6f0d] Deleting UiAutomator2 session
[f34e677f][ADB] Running '/Users/meryl.gregori/Library/Android/sdk/platform-tools/adb -P 5037 -s 9a424a1b shell dumpsys activity services io.appium.settings/.recorder.RecorderService'
[f34e677f][ADB] Running '/Users/meryl.gregori/Library/Android/sdk/platform-tools/adb -P 5037 -s 9a424a1b shell am force-stop com.paymaya.staging'
[HTTP] <-- POST /session - - ms - - 
[f34e677f][AndroidUiautomator2Driver@6f0d] Restoring hidden api policy to the device default configuration
[f34e677f][ADB] Running '/Users/meryl.gregori/Library/Android/sdk/platform-tools/adb -P 5037 -s 9a424a1b shell 'settings delete global hidden_api_policy_pre_p_apps;settings delete global hidden_api_policy_p_apps;settings delete global hidden_api_policy''
[f34e677f][AppiumDriver@30d3] Event 'newSessionStarted' logged at 1734076772553 (15:59:32 GMT+0800 (Philippine Standard Time))
[f34e677f][AppiumDriver@30d3] Encountered internal error running command: Error executing adbExec. Original error: 'Command '/Users/meryl.gregori/Library/Android/sdk/platform-tools/adb -P 5037 -s 9a424a1b install -g /Users/meryl.gregori/.appium/node_modules/appium-uiautomator2-driver/node_modules/io.appium.settings/apks/settings_apk-debug.apk' timed out after 40000ms'. Try to increase the 40000ms adb execution timeout represented by 'androidInstallTimeout' capability
Error: Command '/Users/meryl.gregori/Library/Android/sdk/platform-tools/adb -P 5037 -s 9a424a1b install -g /Users/meryl.gregori/.appium/node_modules/appium-uiautomator2-driver/node_modules/io.appium.settings/apks/settings_apk-debug.apk' timed out after 40000ms
    at Timeout._onTimeout (/Users/meryl.gregori/.appium/node_modules/appium-uiautomator2-driver/node_modules/teen_process/lib/exec.js:122:19)
    at listOnTimeout (node:internal/timers:564:17)
    at processTimers (node:internal/timers:507:7)
`

Appium version: 2.13.1
uiAutomator2 version: 3.9.3
Appium Inspector version: 2024.12.1
Device: Redmi 10C - Android 13

Would also like to ask if slow internet connection attributes as to why appium inspector gets a request timeout issue? Thanks in advance for the help!

**Comment by KazuCocoa** [comment]
> rror: Command '/Users/meryl.gregori/Library/Android/sdk/platform-tools/adb -P 5037 -s 9a424a1b install -g /Users/meryl.gregori/.appium/node_modules/appium-uiautomator2-driver/node_modules/io.appium.settings/apks/settings_apk-debug.apk' timed out after 40000ms at Timeout

It looks like the session preparation to install necessary apks failed as a timeout. You could try out the same command by manual to see how usually it takes. Some devices may have config to ask app installation via `adb` as a popup and it blocks the installation. Then, you may need to modify the device's developer option to skip app validation. It depends on the device kind.
